traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java

```diff
@@ -559,9 +560,17 @@ public HTTPRouteResult multiRoute(final HTTPRequest request, final Track track)
 		return routeResult;
 	}
 
-	public String buildPatternBasedHashString(final DeliveryService deliveryService, final String requestPath) {
+	public String buildPatternBasedHashString(final DeliveryService deliveryService, final HTTPRequest request) {
+		final String requestPath = request.getPath();
 		if (deliveryService.getConsistentHashRegex() != null && !deliveryService.getConsistentHashRegex().isEmpty() && !requestPath.isEmpty()) {
-			return buildPatternBasedHashString(deliveryService.getConsistentHashRegex(), requestPath);
+			final StringBuilder hashString = new StringBuilder(request.getPath());
+			for (String qkey : request.getQueryString().split("&")) {
+				qkey = qkey.split("=")[0];
+				if (deliveryService.consistentHashQueryParams.contains(qkey)) {
+					hashString.append(qkey);
```

Comment:
> _"I thought we needed to include the query param value in the hash string as well, not just the key?"_

... yep. :man_facepalming: oops

> _"Also, we need to sort the query string key/values somehow..."_

I'm pretty sure `SortedSet<E>`s iterate in sorted order. I'd need to check the docs again to be sure (since you've just shaken my faith on that), but that's why I chose that interface.

> _"Also, how should we handle the case of /foo?a=b&a=c -- two keys with the same value? Just include both key/value pairs in the hash string? Later values overwrite earlier values? I'm thinking the former, which is along the lines of what this code is already doing."_

iirc we (@mitchell852, @dangogh, and myself) had decided on the former. Seems the most proper handling, IMO. Lemme know if I'm mis-remembering, guys.

Reply:
> I'm pretty sure SortedSet<E>s iterate in sorted order. I'd need to check the docs again to be sure (since you've just shaken my faith on that), but that's why I chose that interface.

Here we are iterating over the tokens of the query string, not the `deliveryService.consistentHashQueryParams` SortedSet, so I think we have to sort the tokens before iterating over them and adding them to the hashString.

Reply:
What's wrong with me?